project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/ontology/TIBTerminologyServiceIntegration.java

```diff
@@ -0,0 +1,174 @@
+    ONTOLOGIES_WHITELIST.add("envo"); // Environmental Factor Ontology
+  }
+
+  private final URI selectEndpointAbsoluteUrl;
+  private final URI searchEndpointAbsoluteUrl;
+
+  @Autowired
+  public TIBTerminologyServiceIntegration(
+      @Value("${tib.terminology.service.endpoint.select}") String selectEndpoint,
+      @Value("${tib.terminology.service.endpoint.search}") String searchEndpoint,
+      @Value("${tib.terminology.service.api.url}") String tibApiUrl) {
+    this.selectEndpointAbsoluteUrl = URI.create(tibApiUrl).resolve(selectEndpoint);
+    this.searchEndpointAbsoluteUrl = URI.create(tibApiUrl).resolve(searchEndpoint);
+  }
+
+  private static OntologyClass convert(TibTerm term) {
+    return new OntologyClass(term.ontologyPrefix, "", term.iri, term.label, term.shortForm,
+        term.getDescription().orElse(""), "");
+  }
+
+  private static String createOntologyFilterQueryParameter() {
+    return String.join(",", ONTOLOGIES_WHITELIST);
+  }
+
+  private static HttpClient httpClient() {
+    return HttpClient.newBuilder().version(Version.HTTP_2)
+        .followRedirects(Redirect.NORMAL).connectTimeout(
+            Duration.ofSeconds(10)).build();
+  }
+
+  @Override
+  public List<OntologyClass> query(String searchTerm, int offset, int limit) {
+    try {
+      List<TibTerm> result = select(searchTerm, offset, limit);
+      return result.stream().map(TIBTerminologyServiceIntegration::convert).toList();
+    } catch (IOException | InterruptedException e) {
+      log.error("TIB Service search failed. ", e);
+      throw new ApplicationException("Query failed. Please try again.");
+    }
+  }
+
+  @Override
+  public Optional<OntologyClass> searchByCurie(String curie) {
+    try {
+      List<TibTerm> result = searchByOboId(curie, 0, 10);
+      if (result.isEmpty()) {
+        return Optional.empty();
+      }
+      return Optional.of(
+          result.stream().map(TIBTerminologyServiceIntegration::convert).toList().get(0));
+    } catch (IOException | InterruptedException e) {
+      log.error("TIB Service search failed. ", e);
+      throw new ApplicationException("Query failed. Please try again.");
+    }
+  }
+
+  @Override
+  public List<OntologyClass> search(String searchTerm, int offset, int limit) {
+    try {
+      List<TibTerm> result = fullSearch(searchTerm, offset, limit);
+      return result.stream().map(TIBTerminologyServiceIntegration::convert).toList();
+    } catch (IOException | InterruptedException e) {
+      log.error("TIB Service search failed. ", e);
+      throw new ApplicationException("Query failed. Please try again.");
+    }
+  }
+
+  private List<TibTerm> fullSearch(String searchTerm, int offset, int limit)
+      throws IOException, InterruptedException, ApplicationException {
+    if (searchTerm.isBlank()) { // avoid unnecessary API calls
+      return List.of();
+    }
+    HttpRequest termSelectQuery = HttpRequest.newBuilder().uri(URI.create(
```

Comment:
For URL encoding please have a look at https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/net/URLEncoder.html